### PR TITLE
feat(lbaas): private network support

### PR DESCRIPTION
**Pull request by peknur**
Adds support for load balancers listening private networks.  Built on top of [upcloud-go-api](https://github.com/UpCloudLtd/upcloud-go-api) PR [#182](https://github.com/UpCloudLtd/upcloud-go-api/pull/182)

Bumped Go version to 1.8 and Terraform SDK to v2.24.0.

### CODE REVIEW

**Review by nate-droid — APPROVED** [review]

File: examples/resources/upcloud_loadbalancer/resource.tf [hunk 1]
```diff
@@ -19,4 +19,15 @@ resource "upcloud_loadbalancer" "lb" {
   plan              = "development"
   zone              = var.lb_zone
   network           = resource.upcloud_network.lb_network.id
```
Comment (ka-myl) [inline examples/resources/upcloud_loadbalancer/resource.tf:21]: This line should probably be removed from the example

File: internal/service/loadbalancer/loadbalancer.go [hunk 2]
```diff
@@ -42,11 +45,16 @@ func ResourceLoadBalancer() *schema.Resource {
 				Type:        schema.TypeString,
 				Required:    true,
 			},
-			"network": {
-				Description: "Private network UUID where traffic will be routed. Must reside in load balancer zone.",
-				Type:        schema.TypeString,
-				ForceNew:    true,
-				Required:    true,
+			"networks": {
```
Comment (ka-myl) [inline internal/service/loadbalancer/loadbalancer.go:48]: I would use a singular form here, i.e. `network`. Singular form is a bit nicer from a user perspective, for me having to use a plural form to define a single network is a bit confusing. Not a blocker

File: internal/service/loadbalancer/loadbalancer.go [hunk 3]
```diff
@@ -89,17 +97,144 @@ func ResourceLoadBalancer() *schema.Resource {
...
+		"network": {
+			Description: "Private network UUID. Required for private networks and must reside in loadbalancer zone. For public network the field should be omitted.",
+			Type:        schema.TypeString,
+			ForceNew:    true,
+			Optional:    true,
+		},
+		"dns_name": {
+			Description: "DNS name of the load balancer network",
+			Type:        schema.TypeString,
+			ForceNew:    true,
```
Comment (ka-myl) [inline internal/service/loadbalancer/loadbalancer.go:171]: Probably not needed, as this is a computed only attribute

**Review by ka-myl — APPROVED** [review]
LGTM. Tested locally and it works nicely. Left some minor comments, but those are just points to consider, no blockers!

File: internal/service/loadbalancer/loadbalancer.go (hunk 3 above)
Comment (peknur) [reply internal/service/loadbalancer/loadbalancer.go:171]: good catch, I'll fix that.

File: internal/service/loadbalancer/loadbalancer.go (hunk 2 above)
Comment (peknur) [reply internal/service/loadbalancer/loadbalancer.go:48]: That is true. I was thinking to use singular form but examples that I viewed all used plural naming and `network` is still reserved to old field so I decided to go with plural.

File: examples/resources/upcloud_loadbalancer/resource.tf (hunk 1 above)
Comment (peknur) [reply examples/resources/upcloud_loadbalancer/resource.tf]: Yes it should :) It's removed now.

**Pull request merged**